[RISCV][GISel] RegBank select and instruction select for vector G_ADD, G_SUB

jiahanxie353 opened the pull request:
RegisterBank Selection for scalable vector G_ADD and G_SUB by creating new mappings for different types of vector register banks.
Then implement Instruction Selection for the same operations by choosing the correct RISC-V vector register class.

jiahanxie353 commented:
okay, there's nothing showing up under "Reviewers" to request your review, so I'll just @michaelmaitland @topperc

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:
```diff
@@ -240,6 +254,10 @@ RISCVRegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
 
+  unsigned VRBSize = getMaximumSize(RISCV::VRBRegBankID);
+  const ValueMapping *VRBValueMapping =
+      &RISCV::ValueMappings[VRBSize == 64 ? RISCV::VRB64Idx : RISCV::VRB32Idx];
```
The size for VRB bank shouldn't change. We only have to do this for GPR because of RV32 vs RV64.

jiahanxie353 replied:
for VRB, do we always have size == 64?

michaelmaitland replied:
It looks like in `build/lib/Target/RISCV/RISCVGenRegisterBank.inc` that the `Sizes` array has 512 for both RV32 and RV64 for VRB.

topperc replied:
512 is the size of LMUL=8. we should have 4 different sizes in the mapping for the 4 different whole LMUL register sizes. We should pick based on the known minimum size of the type. 512, 256, 128, and 64. Anything less than 64 should use the 64.

jiahanxie353 replied:
Actually I'm a bit lost right now. Can I check my understanding?

1. 
> The size for VRB bank shouldn't change.

Does this refer to `unsigned VRBSize` never change?

2. Based on my understanding of your description, in [`PartMappings`](https://github.com/llvm/llvm-project/blob/c556c9c3792a322dfc3a9c60be9d7d32cf335235/llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp#L27), I'll probably have:
```C++
    {0, 64, VRBRegBank}, // <= 64
    {0, 64, VRBRegBank}, {64, 128, VRBRegBank}, // 128, LMUL=2
    {0, 64, VRBRegBank}, {64, 128, VRBRegBank}, {128, 192, VRBRegBank}, {192, 256, VRBRegBank}, // 256, LMUL=4
    {0, 64, VRBRegBank} .... // 512, LMUL=8
```
?
And that I'll only have `PMI_VRB64` in [`PartialMappingIdx`](https://github.com/llvm/llvm-project/blob/c556c9c3792a322dfc3a9c60be9d7d32cf335235/llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp#L34) since the size of VRB doesn't change?
And in [`ValueMapping`](https://github.com/llvm/llvm-project/blob/c556c9c3792a322dfc3a9c60be9d7d32cf335235/llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp#L41), I'll probably have:
```C++
{&PartMappings[PMI_VRB64], 1},
{&PartMappings[PMI_VB64], 2},
{&PartMappings[PMI_VRB64], 4},
{&PartMappings[PMI_VRB64], 8}
```
?

michaelmaitland commented on llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:
```diff
@@ -269,6 +287,9 @@ RISCVRegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
   case TargetOpcode::G_SEXTLOAD:
   case TargetOpcode::G_ZEXTLOAD:
+    if (MRI.getType(MI.getOperand(0).getReg()).isVector())
+      return getInstructionMapping(DefaultMappingID, /*Cost=*/1,
```
This does regbankselect for many more opcodes than just G_ADD and G_SUB. Can you please change to only map to vector for the instructions you mention in PR title/description + include regbankselect tests for.

michaelmaitland commented:
I checked out your code and I am not seeing early exit before selectImpl for the G_ADD. I only see the COPY and PseudoRET instructions entering into the if (!MI.isPreISelOpcode() || Opc == TargetOpcode::G_PHI) {...} block

michaelmaitland commented:
> > I checked out your code and I am not seeing early exit before selectImpl for the G_ADD. I only see the COPY and PseudoRET instructions entering into the `if (!MI.isPreISelOpcode() || Opc == TargetOpcode::G_PHI) {...}` block.
> 
> did you pick the vector legalizer commits from my other PR?

Yes

jiahanxie353 commented:
Hi,
I'm stuck on [this part](https://github.com/llvm/llvm-project/blob/21b986637b950bb1762a38201223d62c4bca0dce/llvm/lib/CodeGen/GlobalISel/InstructionSelect.cpp#L280C25-L280C71), where it tries to compare `Ty.getSizeInBits()`, which is a `vscale x s32` and `TRI.getRegSizeInBits(*RC)`, which is an `int 64`.
It seems like there is no `>` overloaded to compare these two types.
Is this issue expected?

michaelmaitland commented:
> Hi, I'm stuck on [this part](https://github.com/llvm/llvm-project/blob/21b986637b950bb1762a38201223d62c4bca0dce/llvm/lib/CodeGen/GlobalISel/InstructionSelect.cpp#L280C25-L280C71), where it tries to compare `Ty.getSizeInBits()`, which is a `vscale x s32` and `TRI.getRegSizeInBits(*RC)`, which is an `int 64`. It seems like there is no `>` overloaded to compare these two types. Is this issue expected?

Before the addition of scalable vectors, the size of scalars and fixed vectors is always known. For example an i32 is 32 bits, a float is 32 bits, and 4 x i32 is 4 * 32 bits. With the addition of scalable vectors, we do not know the size at compile time. The size of a type like `vscale x 1 x i32` depends on the value of `vscale`, which is a runtime constant, which makes the number of bits of that type unknown at compile time.

Instead of returning `unsigned` for the size of the type, we must return `TypeSize`, which is capable of representing that the size is scalable. Take a look at [this](f219e03f3c9bea4220925838c3d57d5a993d4d7a) patch which I added in order to make lowerFormalArguments work. You will likely need to make some unsigned -> TypeSize changes. For the exact case you show here, you probably want to take advantage of the `TypeSize::isKnownLT` function.

jiahanxie353 commented:
> > Hi, I'm stuck on [this part](https://github.com/llvm/llvm-project/blob/21b986637b950bb1762a38201223d62c4bca0dce/llvm/lib/CodeGen/GlobalISel/InstructionSelect.cpp#L280C25-L280C71), where it tries to compare `Ty.getSizeInBits()`, which is a `vscale x s32` and `TRI.getRegSizeInBits(*RC)`, which is an `int 64`. It seems like there is no `>` overloaded to compare these two types. Is this issue expected?
> 
> Before the addition of scalable vectors, the size of scalars and fixed vectors is always known. For example an i32 is 32 bits, a float is 32 bits, and 4 x i32 is 4 * 32 bits. With the addition of scalable vectors, we do not know the size at compile time. The size of a type like `vscale x 1 x i32` depends on the value of `vscale`, which is a runtime constant, which makes the number of bits of that type unknown at compile time.
> 
> Instead of returning `unsigned` for the size of the type, we must return `TypeSize`, which is capable of representing that the size is scalable. Take a look at [this](f219e03f3c9bea4220925838c3d57d5a993d4d7a) patch which I added in order to make lowerFormalArguments work. You will likely need to make some unsigned -> TypeSize changes. For the exact case you show here, you probably want to take advantage of the `TypeSize::isKnownLT` function.

Thanks!
Seems like the link to the lowerFormalArguments patch is not working? Can you send again?

michaelmaitland commented:
> Thanks! Seems like the link to the lowerFormalArguments patch is not working? Can you send again?
https://github.com/llvm/llvm-project/commit/ac4ff6168ab803db282f56af05270faf916f22e2

jiahanxie353 committed:
use isKnownGT to compare LLT and TargetRegister

jiahanxie353 commented:
To implement `RISCVInstructionSelector::getRegClassForTypeOnBank`, I'm creating a mapping function that takes in an `LLT` and output `RegClass`.

Let's take `vscale x 1 x s32` as an example `LLT`, the expected `RegClass` is `RISCV::VRRegClassID` (the reason being we treat `LMUL` <= 1 all as `LMUL`=1 since they have similar behavior, if I remember our discussion from yesterday correctly? @michaelmaitland ).

I was trying to use `getElementType` to extract `s32`, and using `getNumElements()` to extract that `1`, which raises an exception. And I believe it's because `getNumElements()` extracts `vscale` and `1` at the same time. Since `vscale` is unknown at compile time, so it can't calculate the exact `NumElements`.

I'm wondering if I'm on the right track, and is there a generic function to extract `n` out of `vscale x n x m`?

michaelmaitland commented:
> is there a generic function to extract n out of vscale x n x m?

https://llvm.org/doxygen/classllvm_1_1details_1_1FixedOrScalableQuantity.html#ac4ab9dd9440c55bee1aa4a1195cee759

jiahanxie353 committed:
getRegClassForTypeOnBank for vector register bank

jiahanxie353 committed:
vector instr sel for add and sub

michaelmaitland commented:
> Seems like I don't need to implement all the ValueMapping for some reason

It worked because the test cases have `regBankSelected: true` and have `vrb` filled in for the vector instructions: 
```
%2:vrb(<vscale x 1 x s8>) = G_ADD %0, %1
   ^^^
```

If you run the full gisel pipeline on a llvm program or `-run-pass=instruction-select` on MIR that just went through the IR translator, regbankselection will fail.

topperc commented:
> I don't understand the `-1, 3 /* e8 */, 3` part in `; RV32I-NEXT: [[PseudoVADD_VV_MF8_:%[0-9]+]]:vr = PseudoVADD_VV_MF8 [[DEF]], [[COPY]], [[COPY1]], -1, 3 /* e8 */, 3 /* ta, ma */`.

The -1 is the VL operand where -1 means VLMAX.

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -785,6 +785,22 @@ const TargetRegisterClass *RISCVInstructionSelector::getRegClassForTypeOnBank(
   // TODO: Non-GPR register classes.
+
+  if (RB.getID() == RISCV::VRBRegBankID) {
+    if (Ty.getSizeInBits().getKnownMinValue() <= 64) {
```
Drop curly braces around single lines.

jiahanxie353 committed:
getInstrMapping for vector add and sub; regbankselect test cases for vector add and sub

jiahanxie353 commented:
What is the `instruction-select` supposed to generate? Is it  supposed to stop and generate vector instructions pseudos, like what I have right now in the test files; or do we want actual RISC-V `vsetvli` and `vadd` instructions after the `instruction-select` pass? 
I think it's the former case and the transformation from pseudos to actual assembly instructions is taken care of by the rest of llvm infrastructure and we don't need to explicitly support in GISel?

michaelmaitland commented:
> I think it's the former case and the transformation from pseudos to actual assembly instructions is taken care of by the rest of llvm infrastructure and we don't need to explicitly support in GISel?

This. 

Instruction selection is responsible for lowering into target specific MIR (`MachineInstr`s). The RISCVInsertVSETVLI pass comes later in the pipeline. There is a RISCVAsmPrinter pass that will convert `MachineInstr` to `MCInst` (`RISCVAsmPrinter::lowerToMCInst`) which is the representation of assembly.

michaelmaitland commented on llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:
```diff
@@ -57,6 +60,22 @@ const RegisterBankInfo::ValueMapping ValueMappings[] = {
     {&PartMappings[PMI_FPRB64], 1},
     {&PartMappings[PMI_FPRB64], 1},
     {&PartMappings[PMI_FPRB64], 1},
+    // Maximum 3 VR LMUL=1 operands.
```
```suggestion
    // Maximum 3 VR LMUL={1, MF2, MF4, MF8} operands.
```

michaelmaitland commented on llvm/test/CodeGen/RISCV/GlobalISel/regbankselect/vec-add-sub-rv32.mir:
```diff
@@ -0,0 +1,511 @@
+tracksRegLiveness: true
+body:             |
+  bb.0.entry:
+    liveins: $v10, $v11
```
Be careful which vector registers you are passing as livein. M8 operations will be using `v10, v11, v12, v13, v14,  v15, v16, v17` as a single register grouping. That means that you probably should use the next available register for the next live in.

I am calling out this specific LMUL 8 case in this test, but make sure this holds for all your test cases for their respective LMULs. It could be helpful to see which registers the IR translator for a test that lowers to this MIR test case chooses if you are not sure.

topperc replied:
> Be careful which vector registers you are passing as livein. M8 operations will be using `v10, v11, v12, v13, v14,  v15, v16, v17` as a single register grouping. That means that you probably should use the next available register for the next live in.
> 
> 
> 
> I am calling out this specific LMUL 8 case in this test, but make sure this holds for all your test cases for their respective LMULs. It could be helpful to see which registers the IR translator for a test that lowers to this MIR test case chooses if you are not sure.

Why hand construct the test at all? Write the IR test and use -stop-before to get the MIR.

jiahanxie353 committed:
drop curly braces around single lines

jiahanxie353 committed:
correct vectorized G_ADD, G_SUB instruction selection test cases

jiahanxie353 committed:
partial mapping VRB64 is for LMUL=1, MF2, MF4, and MF8

michaelmaitland commented on llvm/lib/CodeGen/GlobalISel/InstructionSelect.cpp:
```diff
@@ -277,7 +277,7 @@ bool InstructionSelect::runOnMachineFunction(MachineFunction &MF) {
 
     const LLT Ty = MRI.getType(VReg);
-    if (Ty.isValid() && Ty.getSizeInBits() > TRI.getRegSizeInBits(*RC)) {
+    if (Ty.isValid() && TypeSize::isKnownGT(Ty.getSizeInBits(), TRI.getRegSizeInBits(*RC))) {
```
Clang format needed here?

michaelmaitland reviewed (commented):
can you mention regbankselect in the title/description of PR?

jiahanxie353 committed:
clang format

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -785,6 +785,21 @@ const TargetRegisterClass *RISCVInstructionSelector::getRegClassForTypeOnBank(
   }
 
   // TODO: Non-GPR register classes.
```
I think we can delete this TODO

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:
```diff
@@ -25,17 +25,20 @@ namespace llvm {
-    {0, 64, GPRBRegBank},
-    {0, 32, FPRBRegBank},
-    {0, 64, FPRBRegBank},
+    {0, 32, GPRBRegBank}, {0, 64, GPRBRegBank}, {0, 32, FPRBRegBank},
```
Please put these back on single lines regardless of what clang-format says

jiahanxie353 committed:
delete todo

jiahanxie353 committed:
put partial mappings to single lines

jiahanxie353 committed:
regbankselect test cases for vectorized G_ADD and G_SUB

jiahanxie353 committed:
clang format off to put PartMappings in single lines

michaelmaitland reviewed (approved):
LGTM with nit comment that `RISCV` -> `RISC-V` in commit description.
